### PR TITLE
Apply kind code check on exitstat and cmdstat 

### LINKED ISSUE #77990: No checks for the integer arguments of intrinsic "execute_command_line"
Section 16.9.73 says that the `exitstat` argument should have a decimal exponent range of at least nine, and the `cmdstat` argument should have a decimal exponent range of at least four.  But the following program does not emit any compile time error messages:
```
program bug
  implicit none
  integer(kind = 1) :: exitstatvar
  integer(kind = 1) :: cmdstatvar 
  character(len=256) :: msg
  character(len=:), allocatable :: command
  command='echo hello'
  call execute_command_line(command, exitstat=exitstatvar, cmdstat=cmdstatvar)
end program
```
All other compilers I've tried emit error messages for this program.

### PR DESCRIPTION
**Pull request by yiwu0b11**
When testing on gcc, both exitstat and cmdstat must be a kind=4 integer, e.g. DefaultInt. This patch changes the input arg requirement from `AnyInt` to `TypePattern{IntType, KindCode::greaterOrEqualToKind, n}`.

The standard stated in 16.9.73
 - EXITSTAT (optional) shall be a scalar of type integer with a decimal exponent range of at least nine.
 - CMDSTAT (optional) shall be a scalar of type integer with a decimal exponent range of at least four.
 
```fortran
program bug
  implicit none
  integer(kind = 2) :: exitstatvar
  integer(kind = 4) :: cmdstatvar 
  character(len=256) :: msg
  character(len=:), allocatable :: command
  command='echo hello'
  call execute_command_line(command, exitstat=exitstatvar, cmdstat=cmdstatvar)
end program
```
When testing the above program with exitstatvar kind<4, an error would occur:
```
$ ../build-release/bin/flang-new test.f90 
error: Semantic errors in test.f90
./test.f90:8:47: error: Actual argument for 'exitstat=' has bad type or kind 'INTEGER(2)'
    call execute_command_line(command, exitstat=exitstatvar)
```

When testing the above program with exitstatvar kind<2, an error would occur:
```
$ ../build-release/bin/flang-new test.f90 
error: Semantic errors in test.f90
./test.f90:8:47: error: Actual argument for 'cmdstat=' has bad type or kind 'INTEGER(1)'
    call execute_command_line(command, cmdstat=cmdstatvar)
```

Test file for this semantics has been added to `flang/test/Semantic`
Fixes: https://github.com/llvm/llvm-project/issues/77990

### CODE REVIEW

**Review by klausler — CHANGES_REQUESTED** [review]
Absolutely not.  If the standard doesn't restrict the types of these arguments, neither can we.

**Comment by yi-wu-arm** [comment]
> Absolutely not. If the standard doesn't restrict the types of these arguments, neither can we.

Should it be _Defaulnt_ for existstat, an int32_t(kind==4) would cover +-10^9, and an int16_t(kind==2) for +-10^4?

**Comment by klausler** [comment]
exitstat can be any kind >= 4.  cmdstat can be any kind >= 2.

**Comment by yi-wu-arm** [comment]
> exitstat can be any kind >= 4. cmdstat can be any kind >= 2.

Added a `KindCode::greaterAndEqualTo` in intrinsics to allow this kind check.

File: flang/lib/Evaluate/intrinsics.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ static constexpr CategorySet AnyType{IntrinsicType | DerivedType};
 
 ENUM_CLASS(KindCode, none, defaultIntegerKind,
     defaultRealKind, // is also the default COMPLEX kind
-    doublePrecision, defaultCharKind, defaultLogicalKind,
+    doublePrecision, defaultCharKind, defaultLogicalKind, greaterAndEqualToKind,
```
Comment (klausler) [inline flang/lib/Evaluate/intrinsics.cpp:80]: This name is nonsensical.  How can a value be both greater than and equal to another?

Comment (yi-wu-arm) [reply flang/lib/Evaluate/intrinsics.cpp:80]: would `noLessThanKind` be a better name?

Comment (kiranchandramohan) [reply flang/lib/Evaluate/intrinsics.cpp:80]: `greaterOrEqualsKind` is a possibility.

**Comment by kiranchandramohan** [comment]
You need a test for this patch.

File: flang/lib/Evaluate/intrinsics.cpp (hunk 1 above)
Comment (yi-wu-arm) [reply flang/lib/Evaluate/intrinsics.cpp]: Sorry I was blind... Test file `execute_command_line.f90` has been added to `flang/test/Semantics/` testing for errors when less than, no errors when greater than or equal to.

**Review by klausler — COMMENTED** [review]
Is there another patch that adds integer kind handling for these two arguments to the runtime implementation?

**Comment by yi-wu-arm** [comment]
> Is there another patch that adds integer kind handling for these two arguments to the runtime implementation?

Not now, sorry, let me add that right now.

File: flang/runtime/execute.cpp [hunk 2]
```diff
@@ -122,10 +122,22 @@ void RTNAME(ExecuteCommandLine)(const Descriptor &command, bool wait,
 
   if (exitstat) {
     RUNTIME_CHECK(terminator, IsValidIntDescriptor(exitstat));
+    auto exitstatKind{exitstat->type().GetCategoryAndKind()->second};
+    if (exitstatKind < 4) {
```
Comment (klausler) [inline flang/runtime/execute.cpp:126]: The runtime code doesn't need to check these descriptors -- that will have been done statically when the intrinsic procedure reference was analyzed!

What has to be done in the runtime is using those descriptors to store the integer result values properly, and I think that the code already does that.

**Comment by yi-wu-arm** [comment]
> Is there another patch that adds integer kind handling for these two arguments to the runtime implementation?

Thanks for mention the kind code in the runtime! I've added kind code check in runtime and provide test, an incorrect kind code would result in a `terminator.Crash` "exitstat/cmdstat must have an integer kind greater or equal to 4/2 but have: %d". The tests also test the terminator printout.

**Review by klausler — APPROVED** [review]

File: flang/runtime/execute.cpp (hunk 2 above)
Comment (yi-wu-arm) [reply flang/runtime/execute.cpp:126]: Sorry I misunderstood your comment.

**Comment by yi-wu-arm** [comment]
build and pass all test on Windows on Arm MSVC native.

**Pull request merged**